pkg/http/server.go

```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
+	}
+	err := os.MkdirAll(certDirectory, 0644)
+	if err != nil {
+		return err
+	}
+	// Generate account key
+	aKey, err := newKey(path.Join(certDirectory, "account.pem"))
+	if err != nil {
+		return err
+	}
+
+	// Create the ACME client
+	c := &acme.Client{
+		Key:          aKey,
+		DirectoryURL: letsEncryptURL,
+	}
+
+	// Begin with registration
+	log.Infof("Attempting registration...")
+	_, err = c.Register(context.TODO(), nil, func(string) bool { return true })
+	if err != nil {
+		return err
+	}
+	log.Infof("Registration succeeded")
+
+	// Attempt authorization
+	auth, err := c.Authorize(context.TODO(), s.Domain)
+	if err != nil {
+		return err
+	}
+	var challenge *acme.Challenge
+	for _, c := range auth.Challenges {
+		if c.Type == "http-01" {
+			challenge = c
+		}
+	}
+	if challenge == nil {
+		return err
+	}
+
+	// Determine the correct path to listen on
+	cPath := c.HTTP01ChallengePath(challenge.Token)
+	cResponse, err := c.HTTP01ChallengeResponse(challenge.Token)
+	if err != nil {
+		return err
+	}
+
+	// Create a server that responds to the request
+	mux := http.NewServeMux()
+	mux.HandleFunc(cPath, func(w http.ResponseWriter, r *http.Request) {
+		b := []byte(cResponse)
+		w.Header().Set("Content-Length", strconv.Itoa(len(b)))
+		w.WriteHeader(http.StatusOK)
+		w.Write(b)
+	})
+	l, err := net.Listen("tcp", ":80")
+	if err != nil {
+		return err
+	}
+	defer l.Close()
+	go func() {
+		http.Serve(l, mux)
+	}()
+
+	// Perform the challenge
+	log.Print("Performing challenge...")
+	_, err = c.Accept(context.TODO(), challenge)
+	if err != nil {
+		return err
+	}
+
+	// Wait for authorization to complete
+	_, err = c.WaitAuthorization(context.TODO(), auth.URI)
+	if err != nil {
+		return err
+	}
+	log.Infof("Challenge completed")
+
+	// Generate a key for the domain
+	dKey, err := newKey(fmt.Sprintf("%s.key", path.Join(certDirectory, uDomain)))
+	if err != nil {
+		return err
+	}
+
+	// Create the CSR (certificate signing request)
+	csr, err := x509.CreateCertificateRequest(
+		rand.Reader,
+		&x509.CertificateRequest{
+			Subject: pkix.Name{CommonName: s.Domain},
+		},
+		dKey,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Send the CSR and obtain the certificate
+	log.Infof("Signing the certificate")
+	ders, _, err := c.CreateCert(context.TODO(), csr, 90*24*time.Hour, true)
+	if err != nil {
+		return err
+	}
+	log.Infof("Certificate signed!")
+
+	// Write the certificate bundle to disk
+	w, err := os.Create(path.Join(
+		certDirectory, fmt.Sprintf("%s.crt", uDomain),
+	))
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+	for _, der := range ders {
+		err := pem.Encode(w, &pem.Block{
+			Type:  utils.CertType,
+			Bytes: der,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Infof("Complete!")
+
+	s.TLSCertificateKey = fmt.Sprintf("%s.key", path.Join(certDirectory, uDomain))
+	s.TLSCertificate = path.Join(certDirectory, fmt.Sprintf("%s.crt", uDomain))
+	return nil
+}
+
+func newKey(filename string) (crypto.Signer, error) {
```

Comment:
`crypto.go` already generates private key, maybe we could reuse that (also it uses ecdsa, which has the advantage of a smaller key size).